### ISSUE TITLE
Multiple generic typed clients will only get last added HttpClient

### ISSUE THREAD

**Issue by norachuga** [aspnet/HttpClientFactory#151]
Let's say I have a generic class that needs a uniquely configured HttpClient injected into it.

```
public class ApiInvoker<TService>
{
    public HttpClient Client { get; }
    public ApiInvoker(HttpClient client) 
    {
        Client = client; //should be fully configured already
    }
    // Code that uses TService and Client
}
```
In ConfigureServices, I register several ApiInvokers for different services.

```
services.AddHttpClient<ApiInvoker<IServiceA>>(c => { c.BaseAddress = new Uri("http://aa.aa"); });
services.AddHttpClient<ApiInvoker<IServiceB>>(c => { c.BaseAddress = new Uri("http://bb.bb"); });
services.AddHttpClient<ApiInvoker<IServiceC>>(c => { c.BaseAddress = new Uri("http://cc.cc"); });
```

The expected behavior is that, when injected:
ApiInvoker<IServiceA> will have an HttpClient with a BaseAddress of http://aa.aa.
ApiInvoker<IServiceB> will have an HttpClient with a BaseAddress of http://bb.bb.
ApiInvoker<IServiceC> will have an HttpClient with a BaseAddress of http://cc.cc.

Instead, the outcome is:
ApiInvoker<IServiceA> has an HttpClient with a BaseAddress of http://cc.cc.
ApiInvoker<IServiceB> has an HttpClient with a BaseAddress of http://cc.cc.
ApiInvoker<IServiceC> has an HttpClient with a BaseAddress of http://cc.cc.

I believe this issue is caused by this line:
https://github.com/aspnet/HttpClientFactory/blob/8fa9de4a314bd1ac7ae6a7302a4f5ff508a43763/src/Microsoft.Extensions.Http/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs#L200

In this example, the name of these separate Typed Clients are all **ApiInvoker`1**. This causes each subsequent registration to replace the one above it.

There is an overload that will allows a workaround. You can explicitly set a name for typed clients, so that you can also use them as named clients. If you set each of the names uniquely, the typed clients will not replace each other.
```
services.AddHttpClient<ApiInvoker<IServiceA>>("unique name A", c => { ... });
services.AddHttpClient<ApiInvoker<IServiceB>>("unique name B", c => { ... });
services.AddHttpClient<ApiInvoker<IServiceC>>("unique name C", c => { ... });
```

While this works in the short term, this puts the burden on me to set unnecessary names on clients I'll never need to call by name. If name is strictly required under the hood, it should just generate a GUID for the name instead of using the type name. 

Aside from the naming collision it causes, it could potentially be problematic that an HttpClientFactory could then create named clients that were never specified as named clients. I can call CreateClient("ApiInvoker`1") and get a client I never explicitly registered to that name.

Thanks!

**Comment by martincostello** [comment]
I’m not sure the GUID approach would work as the internal plumbing needs to tell `IHttpClientFactory.CreateClient(string)` what client to create, and it won’t know what to call it with if there’s no way to re-derive the name will it?

I think it does maybe look like a bug though that the type parameter type name(s) aren’t considered for generic types.

In the case of your code you could maybe have an extension something like this that did the naming for you by convention. 

```cs
public static IHttpClientBuilder AddHttpClient<T, TClient>(this IServiceCollection services, Action<HttpClient> configure)
    where T : ApiInvoker<TClient>
{
    return services.AddHttpClient<T>(nameof(TClient), configure);
}
```

**Comment by rynowak** [comment]
Thanks for the report. This looks like a legitimate bug that we can fix.

**Comment by rynowak** [comment]
Fix will be included in 2.2.0-preview2
